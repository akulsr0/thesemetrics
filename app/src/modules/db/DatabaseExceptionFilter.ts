import { ArgumentsHost, Catch, ExceptionFilter } from '@nestjs/common';
import { FastifyReply } from 'fastify';
import { EntityNotFoundError } from 'typeorm/error/EntityNotFoundError';

@Catch(EntityNotFoundError)
export class DatabaseExceptionFilter implements ExceptionFilter {
  catch(exception: EntityNotFoundError, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const response = ctx.getResponse<FastifyReply<Response>>();

    response.status(404).send({
      statusCode: 404,
      message: 'Not found',
    });
  }
}
