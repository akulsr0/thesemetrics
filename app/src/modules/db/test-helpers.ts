import { DynamicModule } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import {
  DailyAggregateDevice,
  DailyAggregatePageView,
  DailyAggregateReferrerPageView,
  PageView,
  Project,
  ProjectEvent,
  Team,
  User,
} from '@thesemetrics/database';
import FS from 'fs';
import Formatter from 'sql-formatter';

export interface SqlQuery {
  statement: string;
  parameters?: any[];
}

export interface SqlError extends SqlQuery {
  message: string;
}

export const SQL_ERRORS = '@test/sql-errors';
export const SQL_QUERIES = '@test/sql-queries';
export class TestDatabaseModule {
  static forRoot(): DynamicModule {
    const queries: SqlQuery[] = [];
    const errors: SqlError[] = [];

    return {
      module: TestDatabaseModule,
      imports: [
        TypeOrmModule.forRoot({
          type: 'sqljs',
          logging: true,
          synchronize: false,
          logger: {
            log: () => {},
            logMigration: () => {},
            logQuerySlow: () => {},
            logSchemaBuild: () => {},
            logQueryError: (message, statement, parameters) => errors.push({ message, statement, parameters }),
            logQuery: (statement, parameters) => queries.push({ statement, parameters }),
          },
          entities: [
            User,
            Team,
            Project,
            PageView,
            DailyAggregateDevice,
            DailyAggregatePageView,
            DailyAggregateReferrerPageView,
            ProjectEvent,
          ],
        }),
      ],
      providers: [
        {
          provide: SQL_QUERIES,
          useValue: queries,
        },
        {
          provide: SQL_ERRORS,
          useValue: errors,
        },
      ],
    };
  }
}

function stringify(data: any) {
  switch (typeof data) {
    case 'undefined':
      return '';
    case 'string':
      return `'${data.replace(/'/g, "\\'")}'`;
    case 'object':
      if (data instanceof Date) {
        return `'${data.toISOString()}'`;
      }
    default:
      return JSON.stringify(data);
  }
}

export const QuerySerializer = {
  serialize(queries: SqlQuery[]) {
    let content = '\n';

    Array.from(queries).forEach((query, index) => {
      content += `-- Query ${index + 1}\n`;
      content += Formatter.format(query.statement + ';', {
        language: 'sql',
        params: query.parameters ? query.parameters.map(stringify) : undefined,
      });
      content += '\n\n';
    });

    return content;
  },
  test(value: any) {
    return (
      Array.isArray(value) &&
      value.every((query) => {
        if (typeof query === 'object') {
          if ('statement' in query) {
            return (
              typeof query.statement === 'string' &&
              (typeof query.statement === 'undefined' || Array.isArray(query.parameters))
            );
          }
        }
        return false;
      })
    );
  },
};
