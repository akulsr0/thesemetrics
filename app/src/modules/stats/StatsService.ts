import {
  DailyAggregateDevice,
  DailyAggregatePageView,
  DailyAggregateReferrerPageView,
  Project,
  ReferrerKind,
} from '@/modules/db';
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { MoreThanOrEqual, Repository } from 'typeorm';

@Injectable()
export class StatsService {
  constructor(
    @InjectRepository(DailyAggregateDevice)
    private readonly devices: Repository<DailyAggregateDevice>,
    @InjectRepository(DailyAggregatePageView)
    private readonly pageviews: Repository<DailyAggregatePageView>,
    @InjectRepository(DailyAggregateReferrerPageView)
    private readonly referrers: Repository<DailyAggregateReferrerPageView>
  ) {}

  async getPageViews(project: Project, resource = '*', since = getLastMonth()) {
    return this.pageviews.find({
      where: { project, date: MoreThanOrEqual(since), path: resource },
    });
  }

  async getReferrers(project: Project, kind: ReferrerKind, resource = '*', since = getLastMonth()) {
    return this.referrers.find({
      where: { project, referrerKind: kind, date: MoreThanOrEqual(since), path: resource },
    });
  }

  async getTopResources(project: Project, since = getLastMonth()) {
    const result: Array<{ path: string; count: string }> = await this.pageviews
      .createQueryBuilder()
      .select([])
      .addSelect('"path"')
      .addSelect('sum("count") as "count"')
      .where({ project, date: MoreThanOrEqual(since) })
      .addGroupBy('"path"')
      .addOrderBy('"count"', 'DESC')
      .take(10)
      .execute();

    return result.map(({ path, count }) => ({ path, count: Number(count) }));
  }

  async getDevices(project: Project, since = getLastMonth()) {
    return this.devices.find({
      where: { project, date: MoreThanOrEqual(since) },
    });
  }
}

function getLastMonth() {
  const date = new Date();

  date.setDate(date.getDate() - 30);

  return date;
}
