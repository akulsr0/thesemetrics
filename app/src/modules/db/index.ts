import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import {
  User,
  Team,
  Project,
  PageView,
  DailyAggregateDevice,
  DailyAggregatePageView,
  DailyAggregateReferrerPageView,
  ProjectEvent,
  config,
} from '@thesemetrics/database';
import { APP_FILTER } from '@nestjs/core';
import { DatabaseExceptionFilter } from '@/modules/db/DatabaseExceptionFilter';

export * from '@thesemetrics/database';

@Module({
  imports: [
    TypeOrmModule.forRoot({
      ...config(),
      entities: [
        User,
        Team,
        Project,
        PageView,
        DailyAggregateDevice,
        DailyAggregatePageView,
        DailyAggregateReferrerPageView,
        ProjectEvent,
      ],
      synchronize: false,
    }),
  ],
  providers: [{ provide: APP_FILTER, useClass: DatabaseExceptionFilter }],
})
export default class DatabaseModule {}
