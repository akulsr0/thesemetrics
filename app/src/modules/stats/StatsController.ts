import { ReferrerKind } from '@/modules/db';
import { ProjectService } from '@/modules/project';
import { StatsService } from '@/modules/stats/StatsService';
import { Controller, Get, Query, Res } from '@nestjs/common';
import { FastifyReply } from 'fastify';

@Controller()
export class StatsController {
  constructor(private readonly stats: StatsService, private readonly projects: ProjectService) {}

  @Get('/stats')
  async getStats(
    @Query('domain') domain: string,
    @Query('path') path: string,
    @Res() response: FastifyReply<Response>
  ) {
    if (path !== '*') path = '/' + path;
    const project = await this.projects.findProjectByDomain(domain);
    const devices = await this.stats.getDevices(project);
    const pageviews = await this.stats.getPageViews(project, path);
    const resources = await this.stats.getTopResources(project);
    const referrers = await this.stats.getReferrers(project, ReferrerKind.REFERRER);

    if (!__DEV__) response.header('Cache-Control', `public, max-age=${secondsRemainingToday()}, immutable`);

    return response.type('application/json').send({
      domain: project.domain,
      resource: path,
      devices,
      pageviews,
      resources,
      referrers,
    });
  }
}

function secondsRemainingToday() {
  const now = new Date();
  const eod = new Date();

  eod.setUTCHours(23, 59, 59, 999);

  return (eod.getTime() - now.getTime()) / 1000;
}
