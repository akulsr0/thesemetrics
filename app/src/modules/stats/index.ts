import { Module } from '@nestjs/common';
import DatabaseModule from '@/modules/db';
import ProjectModule from '@/modules/project';
import { TypeOrmModule } from '@nestjs/typeorm';
import { DailyAggregateDevice, DailyAggregatePageView, DailyAggregateReferrerPageView } from '@thesemetrics/database';
import { StatsController } from '@/modules/stats/StatsController';
import { StatsService } from '@/modules/stats/StatsService';

@Module({
  imports: [
    DatabaseModule,
    ProjectModule,
    TypeOrmModule.forFeature([DailyAggregateDevice, DailyAggregatePageView, DailyAggregateReferrerPageView]),
  ],
  providers: [StatsController, StatsService],
  controllers: [StatsController],
})
export default class StatsModule {}
