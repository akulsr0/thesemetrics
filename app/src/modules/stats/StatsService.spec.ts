import { QuerySerializer, SqlQuery, SQL_QUERIES, TestDatabaseModule } from '@/modules/db/test-helpers';
import { StatsService } from '@/modules/stats/StatsService';
import { Test } from '@nestjs/testing';
import { TypeOrmModule } from '@nestjs/typeorm';
import {
  DailyAggregateDevice,
  DailyAggregatePageView,
  DailyAggregateReferrerPageView,
  Project,
  ReferrerKind,
} from '@thesemetrics/database';

describe('StatsService', () => {
  let service: StatsService;
  let queries: SqlQuery[];
  const since = new Date('2020-05-20T00:00:00.000Z');
  const fakeProject: Project = { id: 1 } as any;

  expect.addSnapshotSerializer(QuerySerializer);

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [
        TestDatabaseModule.forRoot(),
        TypeOrmModule.forFeature([DailyAggregateDevice, DailyAggregatePageView, DailyAggregateReferrerPageView]),
      ],
      providers: [StatsService],
    }).compile();

    service = moduleRef.get(StatsService);
    queries = moduleRef.get(SQL_QUERIES);
  });

  beforeEach(() => {
    queries.length = 0;
  });

  describe('getDevices', () => {
    it('should query aggregated devices for the project', async () => {
      await expect(service.getDevices(fakeProject, since)).rejects.toThrow();
      expect(queries).toMatchSnapshot();
    });
  });

  describe('getPageViews', () => {
    it('should query aggregated pageviews for the project', async () => {
      await expect(service.getPageViews(fakeProject, '*', since)).rejects.toThrow();
      expect(queries).toMatchSnapshot();
    });
  });

  describe('getReferrers', () => {
    it('should query aggregated referrers for the project', async () => {
      await expect(service.getReferrers(fakeProject, ReferrerKind.REFERRER, '*', since)).rejects.toThrow();
      expect(queries).toMatchSnapshot();
    });
  });

  describe('getTopResources', () => {
    it('should query top resources for the project', async () => {
      await expect(service.getTopResources(fakeProject, since)).rejects.toThrow();
      expect(queries).toMatchSnapshot();
    });
  });
});
