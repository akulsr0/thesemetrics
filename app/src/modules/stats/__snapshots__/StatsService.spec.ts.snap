// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsService getDevices should query aggregated devices for the project 1`] = `

-- Query 1
SELECT
  "DailyAggregateDevice"."id" AS "DailyAggregateDevice_id",
  "DailyAggregateDevice"."date" AS "DailyAggregateDevice_date",
  "DailyAggregateDevice"."type" AS "DailyAggregateDevice_type",
  "DailyAggregateDevice"."browser" AS "DailyAggregateDevice_browser",
  "DailyAggregateDevice"."browserVersion" AS "DailyAggregateDevice_browserVersion",
  "DailyAggregateDevice"."os" AS "DailyAggregateDevice_os",
  "DailyAggregateDevice"."osVersion" AS "DailyAggregateDevice_osVersion",
  "DailyAggregateDevice"."count" AS "DailyAggregateDevice_count",
  "DailyAggregateDevice"."projectId" AS "DailyAggregateDevice_projectId"
FROM
  "daily_aggregate_device" "DailyAggregateDevice"
WHERE
  "DailyAggregateDevice"."projectId" = 1
  AND "DailyAggregateDevice"."date" >= '2020-05-20T00:00:00.000Z';


`;

exports[`StatsService getPageViews should query aggregated pageviews for the project 1`] = `

-- Query 1
SELECT
  "DailyAggregatePageView"."id" AS "DailyAggregatePageView_id",
  "DailyAggregatePageView"."date" AS "DailyAggregatePageView_date",
  "DailyAggregatePageView"."path" AS "DailyAggregatePageView_path",
  "DailyAggregatePageView"."count" AS "DailyAggregatePageView_count",
  "DailyAggregatePageView"."uniqueCount" AS "DailyAggregatePageView_uniqueCount",
  "DailyAggregatePageView"."projectId" AS "DailyAggregatePageView_projectId"
FROM
  "daily_aggregate_pageview" "DailyAggregatePageView"
WHERE
  "DailyAggregatePageView"."projectId" = 1
  AND "DailyAggregatePageView"."date" >= '2020-05-20T00:00:00.000Z'
  AND "DailyAggregatePageView"."path" = '*';


`;

exports[`StatsService getReferrers should query aggregated referrers for the project 1`] = `

-- Query 1
SELECT
  "DailyAggregateReferrerPageView"."id" AS "DailyAggregateReferrerPageView_id",
  "DailyAggregateReferrerPageView"."date" AS "DailyAggregateReferrerPageView_date",
  "DailyAggregateReferrerPageView"."referrerKind" AS "DailyAggregateReferrerPageView_referrerKind",
  "DailyAggregateReferrerPageView"."referrer" AS "DailyAggregateReferrerPageView_referrer",
  "DailyAggregateReferrerPageView"."path" AS "DailyAggregateReferrerPageView_path",
  "DailyAggregateReferrerPageView"."count" AS "DailyAggregateReferrerPageView_count",
  "DailyAggregateReferrerPageView"."projectId" AS "DailyAggregateReferrerPageView_projectId"
FROM
  "daily_aggregate_referrer_pageview" "DailyAggregateReferrerPageView"
WHERE
  "DailyAggregateReferrerPageView"."projectId" = 1
  AND "DailyAggregateReferrerPageView"."referrerKind" = 'referrer'
  AND "DailyAggregateReferrerPageView"."date" >= '2020-05-20T00:00:00.000Z'
  AND "DailyAggregateReferrerPageView"."path" = '*';


`;

exports[`StatsService getTopResources should query top resources for the project 1`] = `

-- Query 1
SELECT
  "path",
  sum("count") as "count"
FROM
  "daily_aggregate_pageview" "DailyAggregatePageView"
WHERE
  "DailyAggregatePageView"."projectId" = 1
  AND "DailyAggregatePageView"."date" >= '2020-05-20T00:00:00.000Z'
GROUP BY
  "path"
ORDER BY
  "count" DESC
LIMIT
  10;


`;
